sound/soc/sof/loader.c

```diff
@@ -145,20 +147,41 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 		return ret;
 	}
 
-	/*
-	 * now wait for the DSP to boot. There are 3 possible outcomes:
-	 * 1. Boot wait times out indicating FW boot failure.
-	 * 2. FW boots successfully and fw_ready op succeeds.
-	 * 3. FW boots but fw_ready op fails.
-	 */
-	ret = wait_event_timeout(sdev->boot_wait,
-				 sdev->fw_state > SOF_FW_BOOT_IN_PROGRESS,
-				 msecs_to_jiffies(sdev->boot_timeout));
-	if (ret == 0) {
-		snd_sof_dsp_dbg_dump(sdev, "Firmware boot failure due to timeout",
-				     SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX |
-				     SOF_DBG_DUMP_TEXT | SOF_DBG_DUMP_PCI);
-		return -EIO;
+	if (!sdev->host_ipc_master) {
+		/*
+		 * now wait for the DSP to boot. There are 3 possible outcomes:
+		 * 1. Boot wait times out indicating FW boot failure.
+		 * 2. FW boots successfully and fw_ready op succeeds.
+		 * 3. FW boots but fw_ready op fails.
+		 */
+		ret = wait_event_timeout(sdev->boot_wait,
+					 sdev->fw_state > SOF_FW_BOOT_IN_PROGRESS,
+					 msecs_to_jiffies(sdev->boot_timeout));
+		if (ret == 0) {
+			snd_sof_dsp_dbg_dump(sdev, "Firmware boot failure due to timeout",
+					     SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX |
+					     SOF_DBG_DUMP_TEXT | SOF_DBG_DUMP_PCI);
+			return -EIO;
+		}
+	} else {
+		/* initialize IPC reply buffer if need be */
+		if (!sdev->ipc->max_payload_size) {
+			ret = snd_sof_ipc_init_reply_data_buffer(sdev);
+			if (ret < 0)
+				return ret;
+		}
+
+		/* host is IPC master so SOF_IPC_FW_READY has to be
+		 * initialized by the host. The sof_ipc_fw_ready data
+		 * that was previously signaled by a FW-initiated IPC
+		 * will come as a reply to host's IPC.
+		 */
+		hdr.cmd = SOF_IPC_FW_READY;
+		hdr.size = sizeof(reply);
+
+		ret = sof_ipc_tx_message(sdev->ipc, &hdr, hdr.size, &reply, sizeof(reply));
+		if (ret < 0)
+			return ret;
```

Comment:
Intel may have the issue of skipping the download, but on this new Jailhouse implementation there's more. We're also skipping the "DSP" power on itself, and the firmware starts booting outright before SOF even gets the chance to load -- quite possibly multiple seconds, and if we're doing kernel modules can be as long as 30-60 seconds before the SOF module loads.

I believe this flow change is still relevant because of that. Unless Jailhouse provided a way to delay starting the secondary processor running SOF until the kernel module requested it, that is (there apparently is no such way)

Reply:
The major API interface Jailhouse has in our case is that is controlled only from userspace via ioctls. We tried proposing them opening the API to be also called in kernel, with little success. 

https://groups.google.com/g/jailhouse-dev/c/Vaft0VYnLzY/m/A2TYgLUyAwAJ

Reply:
> it's not clear where the fw_state is set in this block? Is the test line 187 even relevant for this case where the firmware is assumed to have always booted?
> 
> FWIW Intel had such platforms where the firmware is loaded already in ROM/flash, so this notion of skipping the download is relevant.

The fw_state is set in `sof_ipc3_get_reply`. IMO yes it's still relevant as there's no guarantee that the FW is up when module is inserted. A very trivial example for this would be the user inserting the `snd_sof_imx93` module before doing the Jailhouse setup.

Reply:
In this setup the firmware will never going to send the FW_READY notification or it is going to be randomly sent too early, thus the host is not going to receive it?
Iow, is this an issue only on first boot of the firmware and later when you do runtime_pm suspend/resume or system suspend/resume there will be notification?

It might be simpler to handle this use case buy using different description of the setup:
`bool query_fw_ready`  or `no_fw_ready_notification` flag for such platforms.
if it is set, skip waiting for the fw_ready and head directly to ""firmware boot complete" state
implement the `sof_ipc_ops.post_fw_boot` for ipc3 and if the flag is set then query the fw_ready.

This way most of the changes will happen in ipc3.c and probably better layered?